packages/amplify_core_plugin_interface/lib/amplify_auth_category.dart

```diff
@@ -14,64 +13,69 @@ class AuthCategory {
+    return _select((it) => it.resendSignUpCode(request: request));
   }
 
   Future<SignInResult> signIn({@required String username, @required String password, SignInOptions options }) {
     var request = SignInRequest(username: username, password: password, options: options);
-    return plugins.length == 1 ? plugins[0].signIn(request: request) : throw(errorMsg);
+    return _select((it) => it.signIn(request: request));
   }
 
   Future<SignInResult> confirmSignIn({@required String confirmationValue, ConfirmSignInOptions options}) {
     var request = ConfirmSignInRequest(confirmationValue: confirmationValue, options: options);
-    return plugins.length == 1 ? plugins[0].confirmSignIn(request: request) : throw(errorMsg);
+    return _select((it) => it.confirmSignIn(request: request));
   }
 
   Future<SignOutResult> signOut({SignOutOptions options}) {
     var request = SignOutRequest(options: options);
-    return plugins.length == 1 ? plugins[0].signOut(request: request) : throw(errorMsg);
+    return _select((it) => it.signOut(request: request));
   }
 
   Future<UpdatePasswordResult> updatePassword({@required String oldPassword, @required String newPassword, PasswordOptions options}) {
     var request = UpdatePasswordRequest(oldPassword: oldPassword, newPassword: newPassword, options: options);
-    return plugins.length == 1 ? plugins[0].updatePassword(request: request) : throw(errorMsg);
+    return _select((it) => it.updatePassword(request: request));
   }
 
   Future<ResetPasswordResult> resetPassword({@required String username, PasswordOptions options}) {
     var request = ResetPasswordRequest(username: username, options: options);
-    return plugins.length == 1 ? plugins[0].resetPassword(request: request) : throw(errorMsg);
+    return _select((it) => it.resetPassword(request: request));
   }
 
   Future<UpdatePasswordResult> confirmPassword({@required String username, @required String newPassword, @required String confirmationCode, PasswordOptions options}) {
     var request = ConfirmPasswordRequest(username: username, newPassword: newPassword, confirmationCode: confirmationCode, options: options);
-    return plugins.length == 1 ? plugins[0].confirmPassword(request: request) : throw(errorMsg);
+    return _select((it) => it.confirmPassword(request: request));
   }
 
   Future<AuthUser> getCurrentUser() {
     var request = AuthUserRequest();
-    return plugins.length == 1 ? plugins[0].getCurrentUser(request: request) : throw(errorMsg);
+    return _select((it) => it.getCurrentUser(request: request));
   }
 
   Future<AuthSession> fetchAuthSession({AuthSessionOptions options}) {
     var request = AuthSessionRequest(options: options);
-    return plugins.length == 1 ? plugins[0].fetchAuthSession(request: request) : throw(errorMsg);
+    return _select((it) => it.fetchAuthSession(request: request));
+  }
+
+  Future<R> _select<R>(FutureOr<R> onValue(AuthPluginInterface plugin)) {
```

Comment:
Is the argument `AuthPluginInterface plugin` in the callback function only for type inference?

Reply:
I believe so, yes. When you do the work to invert the dependency on `core`, and so will have a `Plugin`, this `_select` could become a base method in `Plugin`.

Reply:
sounds good! I'm currently working on the refactor and it's a big dependency mess right now 😞